File: python/second/count_components.py

class Solution:
    # O(V + E) time | O(V) space
    def countComponents(self, n, edges):
        count = n
        unionFind = UnionFind(n)

        for edge in edges:
            if unionFind.union(edge[0], edge[1]):
                count -= 1

        return count
    
class UnionFind(object):
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [1] * n

    def find(self, n):
        p = n
        while self.parent[p] != p:
            self.parent[p] = self.parent[self.parent[p]]
            p = self.parent[p]
        return p

    def union(self, n1, n2):
        p1 = self.find(n1)        
        p2 = self.find(n2)
        
        if p1 == p2:
            return False
        
        if self.rank[p1] >= self.rank[p2]:
            self.parent[p2] = p1
            self.rank[p1] += 1
        else:
            self.parent[p1] = p2
            self.rank[p2] += 1
        return True 
    
solution = Solution()

n = 6
edges = [[0,1], [1,2], [2,3], [4,5]]

print(solution.countComponents(n, edges))

File: src/main/java/august_2025/CountComponents.java
package august_2025;

public class CountComponents {

    public static void main(String[] args) {
        int n = 6;
        int[][] edges = {{0,1}, {1,2}, {2,3}, {4,5}};

        CountComponents countComponents = new CountComponents();
        int result = countComponents.countComponents(n, edges);
        System.out.println(result);

    }

    // O(v + e) time | O(v) space
    public int countComponents(int n, int[][] edges) {

        UnionFind unionFind = new UnionFind(n);
        int count = n;

        for (int[] edge : edges) {
            int n1 = edge[0];
            int n2 = edge[1];
            if (unionFind.union(n1, n2)) {
                count--;
            }
        }
        return count;

    }

    static class UnionFind {
        int[] parent;
        int[] rank;

        public UnionFind(int n) {
            this.parent = new int[n];
            this.rank = new int[n];

            for (int i = 0; i < n; i++) {
                this.parent[i] = i;
                this.rank[i] = 1;
            }
        }

        private int find(int n) {
            int p = n;
            while (p != parent[p]) {
                parent[p] = parent[parent[p]];
                p = parent[p];
            }
            return p;
        }

        public boolean union(int n1, int n2) {
            int p1 = find(n1);
            int p2 = find(n2);

            if (p1 == p2) {
                return false;
            }

            if (this.rank[p1] > this.rank[p2]) {
                this.parent[p2] = p1;
                this.rank[p1]++;
            } else {
                this.parent[p1] = p2;
                this.rank[p2]++;
            }
            return true;
        }
    }

}
